typing/typecore.ml

```diff
@@ -5689,6 +5688,38 @@ let type_clash_of_trace trace =
     | _ -> None
   ))
 
+(** More precise denomination for type errors. Used by messages:
+
+    - "This <denom> ..."
+    - "The <denom> 'foo' ..." *)
+let exp_denom exp =
+  match exp.pexp_desc with
+  | Pexp_constant _ -> "constant"
+  | Pexp_ident _ -> "value"
+  | Pexp_construct _ | Pexp_variant _ -> "constructor"
+  | Pexp_field _ -> "record field"
+  | Pexp_send _ -> "method"
+  | _ -> "expression"
+
+let report_general_this_exp denom ppf = fprintf ppf "This %s" denom
+
+(** Implements the "This expression" message, printing the expression if it
+    should be according to {!Typedtree.exp_is_nominal}. *)
+let report_this_exp denom ppf exp =
+  let denom = match denom with Some d -> d | None -> exp_denom exp in
+  if Pprintast.exp_is_nominal exp then
+    fprintf ppf "The %s '%a'" denom Pprintast.expression exp
```

Comment:
The expression can now be quoted with `Style.as_inline_code`.

Reply:
I changed to the new quoting mechanism and improved quoting for string constants. I also made longer strings not nominal to avoid ridiculously long error messages.